pepper-apis/src/main/java/org/broadinstitute/ddp/service/UserService.java

```diff
@@ -0,0 +1,129 @@
+     *
+     * @param user the user to be deleted
+     */
+    public void deleteUser(Handle handle, User user) throws IOException {
+        Long userId = user.getId();
+        String userGuid = user.getGuid();
+        LOG.info("Deleting user with id={}", userId);
+        // Removing user profile
+        handle.attach(UserProfileDao.class).getUserProfileSql().deleteByUserId(userId);
+        // Removing medical provider
+        handle.attach(JdbiMedicalProvider.class).deleteByUserId(userId);
+        // Removing user enrollment statuses
+        handle.attach(JdbiUserStudyEnrollment.class).deleteByUserId(userId);
+        // Removing user addresses
+        handle.attach(JdbiTempMailAddress.class).deleteTempAddressByParticipantId(userId);
+        JdbiMailAddress mailAddressDao = handle.attach(JdbiMailAddress.class);
+        mailAddressDao.deleteDefaultAddressByParticipantId(userId);
+        mailAddressDao.deleteAddressByParticipantId(userId);
+        // Removing all the answers and activity instances
+        ActivityInstanceDao instanceDao = handle.attach(ActivityInstanceDao.class);
+        Set<Long> instanceIds = instanceDao.findAllInstanceIdsByUserIds(Collections.singleton(userId));
+        LOG.info("Found {} activity instances to delete", instanceIds.size());
+        DBUtils.checkDelete(instanceIds.size(), instanceDao.deleteAllByIds(instanceIds));
+        // Removing counters, events
+        handle.attach(JdbiEventConfigurationOccurrenceCounter.class).deleteAllByParticipantId(userId);
+        JdbiQueuedEvent queueEvent = handle.attach(JdbiQueuedEvent.class);
+        queueEvent.deleteQueuedNotificationSubstitutionsByUserIds(Collections.singleton(userId));
+        queueEvent.deleteQueuedNotificationsByUserIds(Collections.singleton(userId));
+        queueEvent.deleteQueuedEventsByUserIds(Collections.singleton(userId));
+        // Removing governances
+        UserGovernanceDao userGovernanceDao = handle.attach(UserGovernanceDao.class);
+        List<Governance> userGovernances = userGovernanceDao.findGovernancesByParticipantGuid(user.getGuid()).collect(Collectors.toList());
+        Set<String> studyGuids = userGovernances.stream()
+                .map(Governance::getGrantedStudies).flatMap(Collection::stream).map(GrantedStudy::getStudyGuid)
+                .collect(Collectors.toSet());
+        userGovernanceDao.deleteAllGovernancesForProxy(userId);
+        userGovernanceDao.deleteAllGovernancesForParticipant(userId);
+        handle.attach(StudyGovernanceSql.class).deleteAgeUpCandidateByParticipantId(userId);
+        // Removing user itself
+        handle.attach(DataExportDao.class).deleteDataSyncRequestsForUser(userId);
+        handle.attach(JdbiUser.class).deleteAllByIds(Collections.singleton(userId));
+
+        // Cleaning up the elasticsearch data
+        RestHighLevelClient esClient = ElasticsearchServiceUtil.getElasticsearchClient(ConfigManager.getInstance().getConfig());
```

Comment:
Can we pass in this dependency instead of instantiating it here?

Reply:
I'm not sure I get it. Static method `getElasticsearchClient` tries to get a client for given config ES credentials from "cache" if already instantiated, otherwise creates and caches it for the future calls. Did you mean something else?

Reply:
How about we pass in a ES client in the constructor for `UserService`? We'll call `getElasticsearchClient()` in `DataDonationPlatform` where UserService is instantiated. I believe this will make testing a bit easier, so we can pass in a mock to UserService.

Reply:
Since we have to check new route limitations, that would make impossible to bypass ES for tests. I didn't find better solution to have both DeleteUserRouteTest and DataExporterTest working with the same config. Any thoughts?

Reply:
I will still urge us to consider injecting ES client as dependency. For testing, there are a couple things we can do. One thing to consider is to restructure your code to be more test-able. For example, perhaps encapsulate the route limitation logic into a method that's exposed for testing. Then we can instantiate a "route" object in the test and call that method directly for assertions. Like so:

```
public class MyRoute {
  MyError checkLimits() {
    ...
  }
  Object handle(..) {
    var err = checkLimits();
    if (err != null) {
      // halt!
    }
    ...
  }
}

public class MyRouteTest {
  void test() {
    var route = new MyRoute(mockServiceObject);
    ... route.checkLimits();
  }
}
```

`GetDsmParticipantStatusRoute`, `GetStudyPasswordPolicyRoute`, and `InvitationCheckStatusRoute` are some examples that try to do this. See if that helps.

Reply:
That's done